libraries/core_libs/consensus/src/transaction/gas_pricer.cpp

```diff
@@ -0,0 +1,87 @@
+#include "transaction/gas_pricer.hpp"
+
+namespace taraxa {
+
+GasPricer::GasPricer(uint64_t percentile, uint64_t number_of_blocks, std::shared_ptr<DbStorage> db)
+    : percentile_(percentile), price_list_(number_of_blocks) {
+  if (db) {
+    init_daemon_ = std::make_unique<std::thread>([this, db_ = std::move(db)]() { init(std::move(db_)); });
+  }
+}
+
+GasPricer::~GasPricer() {
+  if (init_daemon_) init_daemon_->join();
+}
+
+u256 GasPricer::bid() const {
+  std::shared_lock lock(mutex_);
+  return latest_price_;
+}
+
+void GasPricer::init(std::shared_ptr<DbStorage> db) {
+  const auto last_blk_num =
+      db->lookup_int<EthBlockNumber>(final_chain::DBMetaKeys::LAST_NUMBER, DB::Columns::final_chain_meta);
+  if (!last_blk_num || *last_blk_num == 0) return;
+  auto block_num = *last_blk_num;
+
+  std::unique_lock lock(mutex_);
+
+  while (price_list_.capacity() != price_list_.size() && block_num) {
+    auto trxs = db->getPeriodTransactions(block_num);
+    block_num--;
+
+    // Light node
+    if (!trxs) {
```

Comment:
not sure if this is ideal, what if you are not light node and getPeriodTransactions does not return the txs for some reason ? In case of full node we could require these txs to exist by assert and in case of light node (we will know it is a light node from config) we could check this